packages/python/plotly/plotly/express/_core.py

```diff
@@ -280,17 +290,14 @@ def make_trace_kwargs(args, trace_spec, trace_data, mapping_labels, sizeref):
         attr_label = get_decorated_label(args, attr_value, attr_name)
         if attr_name == "dimensions":
             dims = [
-                (name, column)
-                for (name, column) in trace_data.items()
+                (name, trace_data.get_column(name))
+                for name in trace_data.columns
                 if ((not attr_value) or (name in attr_value))
-                and (
-                    trace_spec.constructor != go.Parcoords
-                    or _is_continuous(args["data_frame"], name)
-                )
+                and (trace_spec.constructor != go.Parcoords or _is_continuous(df, name))
                 and (
                     trace_spec.constructor != go.Parcats
                     or (attr_value is not None and name in attr_value)
-                    or len(args["data_frame"][name].unique())
+                    or nw.to_py_scalar(df.get_column(name).n_unique())
```

Comment:
@FBruzzesi Is this call to `to_py_scalar` needed? As far as I can tell, `.n_unique()` just returns an ordinary integer.

`.to_py_scalar()` also appears to be very slow in the profiling I've been doing for pie charts, it's taking a significant percentage of the chart generation time.

Reply:
That's to deal with pyarrow, cudf and other non native objects.

@MarcoGorelli I think we should first check for scalar type and have a early return, otherwise start checking non scalar types which require loading libraries. I will make a PR to invert the order

Reply:
@emilykl could you share the script you are using for benchmarking? I am curious to see if changes I am doing have any impact in performances 😇

Reply:
@FBruzzesi Yes, it needs a bit of cleanup first but I'm happy to share. I can confirm that after your and @MarcoGorelli 's recent changes we are seeing equal or better performance compared to `master` for Pandas dataframes for most chart types.